google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/ManagedCloudSdkService.java

```diff
@@ -248,4 +235,58 @@ private void updateStatus(SdkStatus sdkStatus) {
   private void notifyListeners(CloudSdkService sdkService, SdkStatus status) {
     statusUpdateListeners.forEach(listener -> listener.onSdkStatusChange(sdkService, status));
   }
+
+  enum ManagedSdkJobType {
+    INSTALL,
+    UPDATE
+  }
+
+  /**
+   * Managed SDK Job future listener, handles success/error logic, logs errors, shows notifications
+   * to a user, updates SDK service statuses.
+   */
+  private final class ManagedSdkJobListener implements FutureCallback<Path> {
+    private final ManagedSdkJobType jobType;
+
+    private ManagedSdkJobListener(ManagedSdkJobType jobType) {
+      this.jobType = jobType;
+    }
+
+    @Override
+    public void onSuccess(Path path) {
+      logger.info("Managed Google Cloud SDK successfully installed/updated at: " + path);
+
+      updateStatus(SdkStatus.READY);
+
+      ManagedCloudSdkServiceUiPresenter.getInstance().notifyManagedSdkJobSuccess(jobType);
+    }
+
+    @Override
+    public void onFailure(Throwable t) {
+      if (t instanceof InterruptedException) {
+        handleJobCancellation();
+      } else {
+        logger.warn("Error while installing/updating managed Cloud SDK", t);
+        updateStatus(SdkStatus.NOT_AVAILABLE);
+
+        ManagedCloudSdkServiceUiPresenter.getInstance()
+            .notifyManagedSdkJobFailure(jobType, t.toString());
+      }
+    }
+
+    private void handleJobCancellation() {
+      logger.info("Managed Google Cloud SDK install/update cancelled.");
+      // interrupted install means not available, but interrupted update means SDK still available.
+      switch (jobType) {
+        case INSTALL:
+          updateStatus(SdkStatus.NOT_AVAILABLE);
+          break;
+        case UPDATE:
+          updateStatus(SdkStatus.READY);
```

Comment:
can we relatively safely assume that when an update is cancelled, the SDK is still in a "Ready" state (and not corrupted or something)?

Reply:
Good question, let's ask Appu about that.

Reply:
Appu and Mark said most of the time this should true, but on Windows it might break badly in some case. I'm going to add an additional check here for `SDK.isInstalled()` to ensure SDK still in good shape, wdyt?

Reply:
agree.

the next question I guess is: if it breaks badly in windows like you said, are we even in a recoverable state? In other words, can we just call `update` or `install` again later and hope it works?

Reply:
I think it should work - if it doesn't return true in `isInstalled()`, we set status to `NOT_AVAILABLE` and call `install()` next time we need managed SDK - this should erase broken SDK and re-install it fresh.